### DISCUSSION TITLE
Should Renovate use `enqueuePullRequest` when enabling GitHub native auto-merge?

### DISCUSSION THREAD

**Discussion by maxbrunet** [renovatebot/renovate#31315] [Suggest an Idea]
### Tell us more.

I have a repository with 2 rule sets:

1. Require merge queue
2. Require approvals (with private Renovate app added to the bypass list)

When a new update is available, Renovate:

1. Creates a PR
2. Enables GitHub's native auto-merge
3. The PR is not added to the merge queue and never gets merged
4. Every time Renovate re-runs on the repository, it tries its built-in auto-merge and fails with HTTP 405 Method Not Allowed

    ```
    Repository rule violations found
 
    Changes must be made through the merge queue
    ```

I think for this case, Renovate might need to check if merge queue is enabled and use [`enqueuePullRequest`](https://docs.github.com/en/graphql/reference/mutations#enqueuepullrequest) explicitly when enabling auto-merge here:

https://github.com/renovatebot/renovate/blob/85786a62068225dadbe99d08c3c64f8c588a41ec/lib/modules/platform/github/index.ts#L1697-L1700

I have seen this issue being reported multiple times, but never with a solution, oddly `enqueuePullRequest` has never been mentioned (maybe it's a new option?)

**Comment by rarkins** [comment]
I'm pretty sure we use these two features together on this repo without problem. As far as I know GitHub automerge adds it to the merge queue

**Reply by maxbrunet** [reply]
I have seen it it working with the `renovate-approve-bot` (like on this repo) and if I approve the PR, it does go into the merge queue. I suspect the GitHub auto-merge bot (which is mostly hidden from the user perspective) cannot bypass the approval requirement, so the regular flow of auto-merge leads to merge queue cannot happen

**Reply by maxbrunet** [reply]
Authenticated as the Renovate app, I tried to do it manually with:

```bash
gh api graphql -F prNodeID="${PR_NODE_ID}" -f query='
  mutation($prNodeID: ID!) {
    enqueuePullRequest(input: {
      pullRequestId: $prNodeID,
    }) { clientMutationId }
  }'
```

But it is still blocked by the codeowners although Renovate is in the bypass list:

```json
{
  "data": {
    "enqueuePullRequest": null
  },
  "errors": [
    {
      "type": "UNPROCESSABLE",
      "path": [
        "enqueuePullRequest"
      ],
      "locations": [
        {
          "line": 3,
          "column": 5
        }
      ],
      "message": "Pull request Waiting on code owner review from my-org/my-team."
    }
  ]
}
```

If I add Renovate to the bypass list of the merge queue rule set, it can merge directly with its own auto-merge, GitHub's native auto-merge is still blocked. I suspect in both cases of the auto-merge and merge queue, the hidden bots responsible for these features do not honor the bypass list defined in the rule sets

**Reply by rarkins** [reply]
From my past experience you can't bypass the codeowners review no matter what.

**Reply by rarkins** [reply]
Does `enqueuePullRequest` somehow bypass the codeowners problem?

Also I'm wondering if the problem is a little different to what you describe. If you use platform automerge _and_ merge queue then things work fine, because github's automerge knows to use the merge queue. But if you don't use platform automerge then Renovate itself doesn't know how to use merge queue, so will fail?

**Reply by maxbrunet** [reply]
So to sum up, the bypass list of rule sets works for Renovate itself, it does not work for GitHub's auto-merge/merge-queue impersonating/acting on behalf of Renovate. I have reach out the GitHub support, I keep you posted if that goes somewhere.

> Does `enqueuePullRequest` somehow bypass the codeowners problem?

No unfortunately, as per the error I got above 

> But if you don't use platform automerge then Renovate itself doesn't know how to use merge queue, so will fail?

I have not tried that, but I guess it would fail. I do have a hunch it might be a "merge method" Renovate might need to support for GitHub.

`enqueuePullRequest` might also be useful to retry PRs that failed the merge queue and were removed (e.g. flaky tests) without rebasing the branch.

Note: From my testing, `enqueuePullRequest` is not necessarily callable as soon as the PR is opened like auto-merge, all requirements for merging must to be satisfied (e.g. required checks)

**Reply by maxbrunet** [reply]
> There is a Request for Enhancement with our engineering team to allow users to bypass a Ruleset but continue to use the Merge Queue. I have added you as an interested party. I'm unable to tell you if or when this enhancement will be available, but I recommend you keep an eye on the [GitHub changelog](https://github.blog/changelog/) and our social media feeds for the latest announcements about new features and fixes.
>  
> You can also browse our [Public roadmap](https://github.com/orgs/github/projects/4247/views/1) which is a project board that provides a high level overview of the current developments and their planned release schedule

TL;DR: It is not supported at the moment

There might be a need for a separate discussion regarding supporting merge-queue as a merge method for GitHub (when platform auto-merge is disabled) (personally, I do not have a use for it for now)

**Reply by rarkins** [reply]
I created a feature request issue for that: https://github.com/renovatebot/renovate/issues/31525
